### ISSUE TITLE
Question regarding cemconv

### ISSUE THREAD

**Issue by Philla7** [coderbot16/cemconv#1]
Hello @coderbot16 !

I am from Empire Earth Heaven forum and from slack group, Empire Earth Reborn. I didn't know how to contact you until I saw that you are here. Do you still have interest in cemconv? Me and couple of others are still working on game and I extracted many of models from game but they are not fully extracted, some don't have animations, some are missing some parts and other...
If you still have interest or not or at least, reply to me by email, my email is philjohncolt45@gmail.com

Thank you!

**Comment by coderbot16** [comment]
I didn't realize this tool was still in use by anyone! My original purpose was to make a proof of concept for others to build upon, since I had to reverse engineer the assembly of Empire Earth to figure out the model format, and I felt a decoder and encoder was a good way to showcase the format and allow others to build upon it. I am interested in maintaining the cemconv project, but I am currently dedicating my free time towards Minecraft modding. Github is the best way to communicate with me, I've been very active on it in recent times. I will put aside some free time for cemconv in the near future to add some features to it such as extracting individual frames, sub models, as well as adding a tool to view info about CEM files directly.

**Comment by Philla7** [comment]
Yes, it still is hehe. I extracted all models for buildings but they are
missing animations and some parts, I think you said back then that .obj
don't support animations, right? Well, anyway, I want to very thank you for
replying, this means a lot to me and guys who are still in project. Couple
of us even went on another project and that is making a game in Unreal
Engine 4.

With regards,
Philla007

sri, 10. lis 2018. 02:19 coderbot16 <notifications@github.com> je napisao:

**Comment by Philla7** [comment]
By the way, if .OBJ doesn't support animations and other stuff, could you make that .CEM format can be converted to .3DS format if that is possible?

**Comment by coderbot16** [comment]
OBJ was meant as a very simplistic output format for extracting raw model data, and is very lossy in regards to the CEM format, as it does not support animations, tag points, sub models, or any of the features that are closer to the engine. However, implementing FBX or 3DS would take a lot of time, and I think for now it would be beneficial to have an interim solution. My idea is to add a JSON format that is basically a 1:1 representation of the binary CEM structures, allowing direct access for authoring and retrieving CEM information. Then, I'll add tools for slicing out frames and sub models from CEM files. Finally, I will add 3DS or FBX support in the future, since implementing either of those formats could take a good amount of time.

The biggest issue with CEM files is a fact of the time in which they were relevant - they use keyframe animation instead of bone animations, and since bone animations are the dominant type these days, finding a popular format supporting keyframe animation is hard.

**Comment by coderbot16** [comment]
Although, based on your reply, it seems that people are more interested in converting from rather than to CEM files, so I will prioritize those features.

**Comment by Philla7** [comment]
Alright, thank you soon much! By the way are you maybe interested to see or
maybe join us in our seperate project in which we are making game in Unreal
Engine 4? If you want, of course...

čet, 11. lis 2018. 05:39 coderbot16 <notifications@github.com> je napisao:

**Comment by coderbot16** [comment]
I didn't notice your reply, but unfortunately I don't think I would be interested in becoming a part of the project since I just don't play EE anymore. I'm very interested to see what you guys create, but I probably don't have the internal motivation to participate in the project, since I'm currently already invested in the Minecraft modding community.

**Comment by Philla7** [comment]
Alright that's your choice and I respect it... anyway I am looking forward
to new version of cemconv, of course if you will make it when you will have
some time.

Datuma sub, 3. stu 2018. 03:45 coderbot16 <notifications@github.com piše:

**Comment by coderbot16** [comment]
I just pushed a simple update that adds support for extracting individual frames by their index, which you can find on the 0.2.0 release. This enables you to manually combine the frames into morph targets / shape keys in blender. This is just a stopgap until I get FBX support working, since I know that getting animations exported is pretty critical for your project.

**Comment by Philla7** [comment]
Wow, that's incredible! I will test it out today, where can I download it?

Datuma pet, 16. stu 2018. 05:04 coderbot16 <notifications@github.com piše:

**Comment by coderbot16** [comment]
It's on the releases page as release 0.2.0, but here's the direct link: https://github.com/coderbot16/cemconv/releases/download/0.2.0/cemconv-0.2.0-x86_64-pc-windows-msvc.zip

Regarding FBX, I decided to go with exporting COLLADA since it's a text format and easier to work with, like OBJ, but supports much of the same features as FBX. It's also possible to convert it to FBX using blender. I got some basic export work yesterday, but I'm running into a bug with exporting my polygons properly. But, once I get it working, it will support both submodels and animations (maybe even separate materials) due to the flexibility of the COLLADA format!

**Comment by Philla7** [comment]
Alright then, best of luck to you on working with that.

Datuma pet, 16. stu 2018. 16:24 coderbot16 <notifications@github.com piše:

**Comment by coderbot16** [comment]
Experimental COLLADA support is out in the latest version (0.2.1):

![image](https://user-images.githubusercontent.com/22420959/48633973-5f100480-e979-11e8-89a1-f3436fc1f9ed.png)

(I also joined the slack)

**Comment by Philla7** [comment]
Hey man! We didn't hear from you a long time, hope everything is well and
okay...

Philla007

**Comment by coderbot16** [comment]
I'm doing alright, but I've been basically starved of free time since I'm in the junior year of high school, which basically means that there's very little time that I can allocate to my projects. I do feel bad for kinda leaving the community hanging, but the main issue is that it's hard to actually see my code in action because I currently don't have a working copy of Empire Earth ever as I reinstalled my OS 1.5 months ago. Also it kinda complicates things that I'm testing a windows-only game on Linux (wine works but it's buggy). 

I do have some incomplete code for collada imports, but I need to fix it up a bit before it actually works. At some point I will get around to it, since things have cleared up a little bit since the start of the semester.

**Comment by Philla7** [comment]
Alright man, take care of those things first. I totally feel you because I
have many exams this month on college and I have small amount of time for
EE, I mean I check out Slack regularly and answer but don't have time to do
something.

Alright, like I said, take care of things in school first and then other
things.
We will wait for you.

Have a great day and good luck in high school!

Philla007

Datuma pet, 8. velj 2019. 05:58 coderbot16 <notifications@github.com piše:

**Comment by Philla7** [comment]
Also, Zocker1600 from Slack group said that he could help you so shall he
contact you or you will contact him?

And what is your email because this seems that it goes to GitHub...

Datuma pet, 8. velj 2019. 11:07 Filip Rozic <philjohncolt45@gmail.com piše:

**Comment by Philla7** [comment]
Oh yeah I forgot, here is Discord link: https://discord.gg/PDGcD9
We created server there for some things


Datuma pet, 8. velj 2019. 11:32 Filip Rozic <philjohncolt45@gmail.com piše:

**Comment by coderbot16** [comment]
I joined the discord, it seems like I might have some free time this weekend to work. If Zocker wants to help me, that would be nice, in fact the best thing for this project would probably be if someone with the necessary skills and time commitment would be able to take the reins and continue development.

**Comment by ferds814** [comment]
bigger plans for me then just reworking EE. and ill be honest. your converter for cem to obj was FANTASTIC! it definitely saved my project. of course without animations or pre-pointed textures, it'll take time to make my own animations, and attach textures, but that's fine, i feel it'll allow for better improvements. the models mainly help with a BASE plaform to work off of. thank you again.

fyi, it took me 5 hours todo. had to decompile all cem models, then use your program to convert 730 models into obj. about to go and convert those 730 files into stl format so i can use. if you see where im going with this.

sadly, i did run into a SNAG with 11 models being cem v1. and cemconv is telling me to run rust_backtrace. but no information on how to actually run the command.

thank you sooo much! and hopefully ill be done by years end.

edit: EMAIL ferds814@gmail.com -

**Comment by Philla7** [comment]
Ferds, be careful when creating new models because they can have maximum 2442 verts, if your model has over that, the game will crash

**Comment by ferds814** [comment]
Affirmative, seeing as Titan engine is old and built for outdated pc's. That explains why building don't have their back side finished.

If you know about this RUST_BACKTRACE=1. let me know. Its killing me. 

I started a new issue in regards to it.

**Comment by zocker-160** [comment]
> I joined the discord, it seems like I might have some free time this weekend to work. If Zocker wants to help me, that would be nice, in fact the best thing for this project would probably be if someone with the necessary skills and time commitment would be able to take the reins and continue development.

I have completely forgotten about this thread over here, but hey better late than never.
I have continued to develop CEMconv and created some additional tools including a blender addon for importing and exporting CEM files.
It is still WiP, but I'm actively working on it  ;)

link to the new repo:
https://github.com/EE-modders/CEM-tool
